lib/std/Build/Watch.zig

```diff
@@ -237,6 +237,194 @@ const Os = switch (builtin.os.tag) {
+                        },
+                    ),
+                };
+                return result;
+            }
+
+            fn deinit(self: *@This(), gpa: Allocator) void {
+                _ = windows.kernel32.CancelIo(self.handle);
+                windows.CloseHandle(self.getWaitHandle());
+                windows.CloseHandle(self.handle);
+                gpa.destroy(self);
+            }
+        };
+
+        fn markDirtySteps(w: *Watch, gpa: Allocator, dir: *Directory) !bool {
+            var any_dirty = false;
+            const bytes_returned = try windows.GetOverlappedResult(dir.handle, &dir.overlapped, false);
+            if (bytes_returned == 0) {
+                std.log.warn("file system watch queue overflowed; falling back to fstat", .{});
+                markAllFilesDirty(w, gpa);
+                return true;
+            }
+            var file_name_buf: [std.fs.max_path_bytes]u8 = undefined;
+            var notify: *align(1) windows.FILE_NOTIFY_INFORMATION = undefined;
+            var offset: usize = 0;
+            while (true) {
+                notify = @ptrCast(&dir.buffer[offset]);
+                const file_name_field: [*]u16 = @ptrFromInt(@intFromPtr(notify) + @sizeOf(windows.FILE_NOTIFY_INFORMATION));
+                const file_name_len = std.unicode.wtf16LeToWtf8(&file_name_buf, file_name_field[0 .. notify.FileNameLength / 2]);
+                const file_name = file_name_buf[0..file_name_len];
+                if (w.os.handle_table.getIndex(dir.handle)) |reaction_set_i| {
+                    const reaction_set = w.os.handle_table.values()[reaction_set_i];
+                    if (reaction_set.getPtr(".")) |glob_set|
+                        any_dirty = markStepSetDirty(gpa, glob_set, any_dirty);
+                    if (reaction_set.getPtr(file_name)) |step_set| {
+                        any_dirty = markStepSetDirty(gpa, step_set, any_dirty);
+                    }
+                }
+                if (notify.NextEntryOffset == 0)
+                    break;
+
+                offset += notify.NextEntryOffset;
+            }
+
+            dir.readChanges();
+            return any_dirty;
+        }
+
+        fn update(w: *Watch, gpa: Allocator, steps: []const *Step) !void {
+            // Add missing marks and note persisted ones.
+            for (steps) |step| {
+                for (step.inputs.table.keys(), step.inputs.table.values()) |path, *files| {
+                    const reaction_set = rs: {
+                        const gop = try w.dir_table.getOrPut(gpa, path);
+                        if (!gop.found_existing) {
+                            var realpath_buf: [std.fs.max_path_bytes]u8 = undefined;
+                            const realpath = try path.root_dir.handle.realpath(path.sub_path, &realpath_buf);
+                            const realpath_w = try windows.sliceToPrefixedFileW(null, realpath);
+                            const dir_handle = windows.kernel32.CreateFileW(
```

Comment:
Why is this `realpath` call necessary? Why is `CreateFileW` used instead of a `std.fs.Dir`/`std.os.windows` function? My naive suggestion would be to change this so something like:

```zig
const dir_handle = try path.root_dir.handle.openDir(path.sub_path, .{});
```

Relevant issues:
- https://github.com/ziglang/zig/issues/16571
- https://github.com/ziglang/zig/issues/19353
- https://github.com/ziglang/zig/issues/1840

Reply:
- I used `realpath` to retrieve an absolute path. I couldn't figure out a way to pass a sub path and root directory handle to CreateFileW.
- CreateFileW was used because I don't think there is a way with `std.fs.Dir/std.os.windows` to provide these flags: FILE_FLAG_BACKUP_SEMANTICS, FILE_FLAG_OVERLAPPED.
- FILE_FLAG_BACKUP_SEMANTICS is a requirement for [ReadDirectoryChangesW](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-readdirectorychangesw).
- FILE_FLAG_OVERLAPPED is for asynchronous I/O (ReadDirectoryChangesW will trigger the event handle).

Reply:
Extracting the `CreateFileW` call from this PR and running it via [`NtTrace`](https://github.com/rogerorr/NtTrace) gives the following [`NtCreateFile`](https://learn.microsoft.com/en-us/windows/win32/api/winternl/nf-winternl-ntcreatefile) call:

```
NtCreateFile( FileHandle=0x701e7ff900 [0xa0], DesiredAccess=SYNCHRONIZE|GENERIC_READ|0x80, ObjectAttributes="\??\C:\Users\Ryan\Programming\Zig\tmp", IoStatusBlock=0x701e7ff908 [0/1], AllocationSize=null, FileAttributes=0, ShareAccess=7, CreateDisposition=1, CreateOptions=0x4000, EaBuffer=null, EaLength=0 ) => 0
```

The important bit here is the `CreateOptions` value, which only has the `FILE_OPEN_FOR_BACKUP_INTENT` flag set, since AFAICT async is the default (there are explicit `FILE_SYNCHRONOUS_IO_NONALERT` and `FILE_SYNCHRONOUS_IO_ALERT` flags instead for `NtCreateFile`).

You're right that there's no existing `std` API that can be used here:
- `std.fs.Dir.openDir` unconditionally sets the `FILE_SYNCHRONOUS_IO_NONALERT` flag, so it's not eligible.
- `std.os.windows.OpenFile` doesn't allow specifying `CreateOptions` and instead automatically populates it (and I think would generally handle opening directories incorrectly, since it never adds the `FILE_FLAG_BACKUP_SEMANTICS` flag)

The 'quick' fix would be to switch to a manual `NtCreateFile` call to satisfy https://github.com/ziglang/zig/issues/1840, and get rid of the `realpath` call. I believe that would look like this (mostly adapted from `std.fs.Dir.makeOpenDirAccessMaskW`):

```zig
const root_fd = path.root_dir.handle.fd;
const sub_path = path.subPathOrDot();
const sub_path_w = try windows.sliceToPrefixedFileW(root_fd, sub_path);
const path_len_bytes = std.math.cast(u16, sub_path_w.len * 2) orelse return error.NameTooLong;

var nt_name = windows.UNICODE_STRING{
    .Length = @intCast(path_len_bytes),
    .MaximumLength = @intCast(path_len_bytes),
    .Buffer = @constCast(sub_path_w.span().ptr),
};
var attr = windows.OBJECT_ATTRIBUTES{
    .Length = @sizeOf(windows.OBJECT_ATTRIBUTES),
    .RootDirectory = if (std.fs.path.isAbsoluteWindowsW(sub_path_w.span())) null else dir_fd,
    .Attributes = 0, // Note we do not use OBJ_CASE_INSENSITIVE here.
    .ObjectName = &nt_name,
    .SecurityDescriptor = null,
    .SecurityQualityOfService = null,
};
var io: windows.IO_STATUS_BLOCK = undefined;
var dir_handle: windows.HANDLE = undefined;
const rc = windows.ntdll.NtCreateFile(
    &dir_handle,
    windows.SYNCHRONIZE | windows.GENERIC_READ | windows.FILE_LIST_DIRECTORY,
    &attr,
    &io,
    null,
    0,
    windows.FILE_SHARE_READ | windows.FILE_SHARE_WRITE | windows.FILE_SHARE_DELETE,
    windows.FILE_OPEN,
    windows.FILE_DIRECTORY_FILE | windows.FILE_OPEN_FOR_BACKUP_INTENT,
    null,
    0,
);

switch (rc) {
    .SUCCESS => {},
    .OBJECT_NAME_INVALID => return error.BadPathName,
    .OBJECT_NAME_NOT_FOUND => return error.FileNotFound,
    .OBJECT_NAME_COLLISION => return error.PathAlreadyExists,
    .OBJECT_PATH_NOT_FOUND => return error.FileNotFound,
    .NOT_A_DIRECTORY => return error.NotDir,
    // This can happen if the directory has 'List folder contents' permission set to 'Deny'
    .ACCESS_DENIED => return error.AccessDenied,
    .INVALID_PARAMETER => unreachable,
    else => return windows.unexpectedStatus(rc),
}
```

This results in the following `NtCreateFile` call when run via `NtTrace`:

```
NtCreateFile( FileHandle=0x1d8e7ff8c0 [0x9c], DesiredAccess=SYNCHRONIZE|GENERIC_READ|0x1, ObjectAttributes=0x48:"", IoStatusBlock=0x1d8e7ff880 [0/1], AllocationSize=null, FileAttributes=0, ShareAccess=7, CreateDisposition=1, CreateOptions=0x4001, EaBuffer=null, EaLength=0 ) => 0
```

Some notes about the differences to the `CreateFileW` call:

- I specified `FILE_LIST_DIRECTORY` in `DesiredAccess` since the [`ReadDirectoryChangesW` docs mention needing it](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-readdirectorychangesw); it might already be included in `GENERIC_READ` for directories but no harm in specifying it manually
- The `CreateFileW` call set `0x80` in `DesiredAccess` which is `FILE_READ_ATTRIBUTES`, but I don't think that's necessary to set explicitly so I didn't set it
- I added `FILE_DIRECTORY_FILE` to `CreateOptions` which will cause `NtCreateFile` to return an error if a non-directory file is attempted to be opened

I think the longer term fix should probably be improving the `std.os.windows.OpenFile` function to be usable both here and in `std.fs.Dir.openDir`, and then use that in both places.